oauth2l-go/src/oauth2client/oauth2client.go

```diff
@@ -0,0 +1,298 @@
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		panic("Failed to read http body for verification request.")
+	}
+	var token map[string]interface{}
+	if err := json.Unmarshal(body, &token); err != nil {
+		panic("Failed to parse json.\n")
+	}
+
+	return token
+}
+
+// Run the three-legged oauth flow, including a authorize flow and a verify
+// flow.
+func run3LOFlow(secret map[string]interface{}, scopes []string) map[string]interface{} {
+	joinedScopes := scopesToString(scopes)
+
+	// In the authorize flow, user will paste a verification code back to console.
+	code := authorizeFlow(secret, joinedScopes)
+
+	// The verify flow takes in the verification code from authorize flow, sends a
+	// POST request containing the code to fetch oauth token.
+	return verifyFlow(secret, joinedScopes, code)
+}
+
+// Helper struct used in sign JWT
+type SignerOpts struct{}
+
+func (r SignerOpts) HashFunc() crypto.Hash {
+	return crypto.SHA256
+}
+
+// Base 64 encode a block. The output doesn't contain the trailing double equal
+// signs.
+func base64Encode(b []byte) string {
+	return strings.TrimSuffix(base64.URLEncoding.EncodeToString(b), "==")
+}
+
+// Creates a signed JWT token. Used for two-legged oauth token request.
+func createJWT(secret map[string]interface{}, joinedScopes string, signer Signer) string {
+	// A valid JWT has an "iat" timestamp and an "exp" timestamp. Get the current
+	// time to create these timestamps.
+	now := int(time.Now().Unix())
+
+	// Construct the JWT header, which contains the private key id in the service
+	// account secret.
+	header := map[string]string{
+		"typ": "JWT",
+		"alg": "RS256",
+		"kid": secret["private_key_id"].(string),
+	}
+
+	// Construct the JWT payload.
+	payload := map[string]string{
+		"aud":   secret["token_uri"].(string),
+		"scope": joinedScopes,
+		"iat":   strconv.Itoa(now),
+		"exp":   strconv.Itoa(now + 3600),
+		"iss":   secret["client_email"].(string),
+	}
+
+	// Convert the header and payload map to json.
+	headerJson, err := json.Marshal(header)
+	if err != nil {
+		panic("Failed to convert JWT header to json.")
+	}
+	payloadJson, err := json.Marshal(payload)
+	if err != nil {
+		panic("Failed to convert JWT payload to json.")
+	}
+
+	// Convert the header and payload json to base64
+	headerB64 := base64Encode(headerJson)
+	payloadB64 := base64Encode(payloadJson)
+
+	// The first two segments of the JWT are signed. The signature is the third
+	// segment.
+	segments := headerB64 + "." + payloadB64
+
+	// Sign the hash, instead of the actual segments.
+	hash := sha256.New()
+	hash.Write([]byte(segments))
+	hashed := hash.Sum(nil)
+	signerOpts := SignerOpts{}
+	signedBytes, err := signer.Sign(rand.Reader, hashed, crypto.SignerOpts(signerOpts))
+	if err != nil {
+		panic(fmt.Sprintf("Failed to sign JWT: %v", err))
+	}
+
+	// Convert the signature to base64
+	signatureB64 := base64Encode(signedBytes)
+
+	// The final JWT has three segments.
+	return segments + "." + signatureB64
```

Comment:
Merge above few lines.

// Generate the final JWT as "base64(header).base64(payload).base64(signature)".
return segments + "." + base64Encode(signedBytes)